x/mongo/driver/topology/pool.go

```diff
@@ -601,3 +527,392 @@ func (p *pool) clear(serviceID *primitive.ObjectID) {
+	wait := func() (*wantConn, *connection, bool) {
+		p.connsCond.L.Lock()
+		defer p.connsCond.L.Unlock()
+
+		for !condition() {
+			p.connsCond.Wait()
+		}
+
+		if ctx.Err() != nil {
+			return nil, nil, false
+		}
+
+		p.newConnWait.cleanFront()
+		w := p.newConnWait.popFront()
+		if w == nil {
+			return nil, nil, false
+		}
+
+		conn, err := newConnection(p.address, p.connOpts...)
+		if err != nil {
+			w.tryDeliver(nil, err)
+			return nil, nil, false
+		}
+
+		conn.pool = p
+		conn.poolID = atomic.AddUint64(&p.nextID, 1)
+		p.conns[conn.poolID] = conn
+
+		return w, conn, true
+	}
+
+	for ctx.Err() == nil {
+		w, conn, ok := wait()
+		if !ok {
+			continue
+		}
+
+		if p.monitor != nil {
+			p.monitor.Event(&event.PoolEvent{
+				Type:         event.ConnectionCreated,
+				Address:      p.address.String(),
+				ConnectionID: conn.poolID,
+			})
+		}
+
+		conn.connect(context.Background())
+		err := conn.wait()
+		if err != nil {
+			_ = p.removeConnection(conn, event.ReasonConnectionErrored)
+			_ = p.closeConnection(conn)
+			w.tryDeliver(nil, err)
+			continue
+		}
+
+		if p.monitor != nil {
+			p.monitor.Event(&event.PoolEvent{
+				Type:         event.ConnectionReady,
+				Address:      p.address.String(),
+				ConnectionID: conn.poolID,
+			})
+		}
+
+		if w.tryDeliver(conn, nil) {
+			continue
+		}
+
+		_ = p.checkInNoEvent(conn)
+	}
+}
+
+func (p *pool) maintain(ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	ticker := time.NewTicker(p.maintainInterval)
+	defer ticker.Stop()
+
+	// remove removes the *wantConn at index i from the slice and returns the new slice. The order
+	// of the slice is not maintained.
+	remove := func(arr []*wantConn, i int) []*wantConn {
+		end := len(arr) - 1
+		arr[i], arr[end] = arr[end], arr[i]
+		return arr[:end]
+	}
+
+	// removeNotWaiting removes any wantConns that are no longer waiting from given slice of
+	// wantConns. That allows maintain() to use the size of its wantConns slice as an indication of
+	// how many new connection requests are outstanding and subtract that from the number of
+	// connections to ask for when maintaining minPoolSize.
+	removeNotWaiting := func(arr []*wantConn) []*wantConn {
+		for i := len(arr) - 1; i >= 0; i-- {
+			w := arr[i]
+			if !w.waiting() {
+				arr = remove(arr, i)
+			}
+		}
+
+		return arr
+	}
+
+	wantConns := make([]*wantConn, 0, p.minSize)
+	defer func() {
+		for _, w := range wantConns {
+			w.tryDeliver(nil, ErrPoolDisconnected)
+		}
+	}()
+
+	for {
+		p.removePerishedConns()
+
+		// Remove any wantConns that are no longer waiting.
+		wantConns = removeNotWaiting(wantConns)
+
+		// Figure out how many more wantConns we need to satisfy minPoolSize. Assume that the
+		// outstanding wantConns (i.e. the ones that weren't removed from the slice) will all return
+		// connections when they're ready, so only add wantConns to make up the difference.
+		total := p.totalConnectionCount()
+		n := int(p.minSize) - total - len(wantConns)
```

Comment:
Is `int(p.minSize) - total - len(wantConns)` the correct calculation here? IIUC `minPoolSize` is the number of *idle* connections that should exist in the pool and `maxPoolSize` is the total number of connections that should exist altogether (i.e. idle + checked out conns). Consider the following case:

1. `minPoolSize` = 5, `maxPoolSize` = 100
2. 10 connections have been created and are currently checked out for operations (i.e. `totalConnectionCount` returns 10).
3. maintenance runs and has no pending `wantConn`s. It will calculate `n := p.minSize - total - len(wantConns)` -> `n := 5 - 10 - 0` -> `n := -5`. This will lead to no new connections being created, even though we should create 5 more.

I _think_ the answer here is to check how many idle connections are in the pool rather than check how many total connections have been created.